[Woo POS] Coupons: Integrate empty | error | disabled | loading states within the view

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

There are a few changes in this PR to integrate error, disabled, empty, and loading states within the coupons view:

1. Moved coupon disabled and error states from the container state to list state ([Commit](https://github.com/woocommerce/woocommerce-ios/commit/e1388768e6a28304ab4c1f43dac70e0b8c29f9a5))
2. Moved default coupons loading state from container to list ([Commit](https://github.com/woocommerce/woocommerce-ios/commit/9c9d329ba694fa700474dd0f96b3fdd0fcb14815)]
3. Moved empty state from cointainer to item list. Both product and coupon empty states will now appear in the list rather than full screen. ([1](https://github.com/woocommerce/woocommerce-ios/commit/9003b3f9bf54e861375ffd65b65b969459526b9f), [2](https://github.com/woocommerce/woocommerce-ios/commit/f98efc84d91869196361dcb0240c6e3b3ce10daa), [3](https://github.com/woocommerce/woocommerce-ios/commit/649e102356975ab6ef77021b14ff8dc8c464dd77), [4](https://github.com/woocommerce/woocommerce-ios/commit/3e57edfa2248bd565ca318e395dbd2b11e9b74f9), [5](https://github.com/woocommerce/woocommerce-ios/commit/48569a5365e4f1f0c588f64ba011ba896e201d92), [6](https://github.com/woocommerce/woocommerce-ios/commit/e3de87f00fa328ede776936114afeb5d19aa808f)
4. Updated tests ([Commit](https://github.com/woocommerce/woocommerce-ios/commit/5b651ca386e40bd4dcd2f97ebaae114512c75b9a))
5. Moved dummy Products / Coupons switching view to a header to more easily evaluate UI of empty and error states ([Commit](https://github.com/woocommerce/woocommerce-ios/commit/7cb54c731f07ecc6efa9d3d250ba97e7a8835af8))

A few unrelated changes:
- [Launch coupon creation from an empty view](https://github.com/woocommerce/woocommerce-ios/commit/12e00227eadc70efcb24693cb104131f367a9c8d)
- [Display a coupons loading error if there's no network when loading coupons](https://github.com/woocommerce/woocommerce-ios/commit/651efa7b99f1a4a766ff747ca38e02aa05cb64db)

### Out of Scope

We will still need to adjust the copy for generic error states. There are still tasks to update designs and copy for Empty and Error views. We also have a task to properly update Item Rows for coupons and its loading state.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Enable `enableCouponsInPointOfSale` feature flag

There are a few different flows to test:
- Empty products state. (e.g return empty state from `PointOfSaleItemService`)
- Empty coupons state. (e.g return empty state from `PointOfSaleCouponService`)
- Disabled coupons state (disable WooCommerce -> Settings coupons)
- Loading state (switch to coupons view)
- Generic coupons error state (disable network when switching to coupons)

All of these states should appear within the coupons list. Error and loading state for products still appear full screen.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I will still want to run a few regression tests before merging:

- [x] Do regression testing for products and product variations when feature flag is disabled

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Generic Error

https://github.com/user-attachments/assets/625e44ca-6172-4ae7-a537-fb856daba521

### Empty Products and Coupons + Loading


https://github.com/user-attachments/assets/6db8a67f-cfb2-4fb9-9973-d14c43d578b1

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.
This pull request includes several changes to improve the handling of view states and error states in the Point of Sale (POS) system. The most significant changes involve the introduction of new view states, refactoring error state handling, and updating UI components to reflect these changes.

staskus committed:
Launch coupon creation from an empty view

staskus committed:
Move Coupon disabled and error states to the list

staskus committed:
Change default coupons loading state to list loading state

Show shimmering coupon rows instead of a large loading view

staskus committed:
Show loading state when enabling coupons

staskus committed:
Display a coupons loading error if there's no network when loading coupons list

checkStoreCouponSettings fails and shows couponsDisabled although there's an unrelated network issue

staskus committed:
Update PointOfSaleItemListEmptyView to support coupons

staskus committed:
Remove empty dashboard state

staskus committed:
Create empty item list state

Empty item list state which is shown within products / coupons section replaces a full-screen container empty state.

staskus committed:
Display emptyView in ItemListView

staskus committed:
Remove noCouponsFound error in PointOfSaleErrorState

staskus committed:
Move dummy products-coupons switcher to the header

It allows to evaluate the look of empty and error state UI properly

staskus committed:
Adjust error and empty views to have button with 50% width

staskus committed:
Move errorView into a separate method

Turn PointOfSaleErrorState static func to static var so they could be used within enum without a warning

staskus committed:
Update POS tests for moving empty state from container to item list

staskus committed:
Test errorOnLoadingCoupons is thrown when retrieving coupon settings fails

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma commented on WooCommerce/Classes/POS/Models/PointOfSaleErrorState.swift:
```diff
@@ -16,55 +16,47 @@ struct PointOfSaleErrorState: Equatable {
     let buttonText: String
 
-    static func errorOnLoadingProducts() -> Self {
+    static var errorOnLoadingProducts: Self {
```
That's neat 💯

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -112,6 +82,7 @@ private extension ItemListView {
     var headerView: some View {
         VStack {
             POSPageHeaderView(title: Localization.title, trailingContent: {
+                temporaryProductsCouponsSwitcher
```
Nice, specially now that we errors views are encapsulated to the ItemList

iamgabrielma reviewed (approved):
Thanks for working on this, view states are nicer to work with and make more sense (no more error for empty coupons):

Tested with on iPad Mini simulator iOS 18.3

...
✅ Generic errors
✅ Empty Products and Coupons + Loading
... 
✅ Empty products state.
✅ Empty coupons state.
✅ Disabled coupons state (disable WooCommerce -> Settings coupons)
✅ Loading state (switch to coupons view) 
✅ Generic coupons error state (disable network when switching to coupons)
...
👉 Regression testing for products and product variations when feature flag is disabled: I ran a quick test for products and variations for empty cases:

![Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-04-10 at 10 08 03](https://github.com/user-attachments/assets/0cf59cd3-a2a3-4fbb-a9bc-0a8b660634b9)
![Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-04-10 at 10 10 05](https://github.com/user-attachments/assets/c21e64cc-fd6d-4c9d-8a52-c5122b27e1d1)

- The empty screen for products doesn't show full-screen anymore, but I think we should be fine with it as is a general improvement derived from the coupons design iteration. It's also the only view across all empty/error states that shows full-screen, no reason to keep it temporarily for longer if the change has already been approved by design.

![Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-04-10 at 10 17 05](https://github.com/user-attachments/assets/5ad04971-3749-4605-8211-2e6a898fa9c4)

- In `PointOfSaleIemListEmptyView` we've modified the "search" icon for a warning icon and the fonts style (ie: title posOnSurfaceVariantHighest to posOnSurface)
![Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-04-10 at 10 21 05](https://github.com/user-attachments/assets/51e749df-afb0-4b17-8b25-ec3559d61e40)

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleItemListEmptyView.swift:
```diff
@@ -5,42 +5,75 @@ struct PointOfSaleItemListEmptyView: View {
-                .foregroundStyle(Color.posOnSurfaceVariantHighest)
-                .font(.posBodyLargeRegular())
-                .padding([.leading, .trailing])
+            VStack(alignment: .center, spacing: POSSpacing.none) {
```
I believe the changes here creates a regression in the empty state when the feature flag is off, I've dropped some images in the response 👍

staskus replied:
Thanks for catching this!

staskus commented:
> The empty screen for products doesn't show full-screen anymore, but I think we should be fine with it as is a general improvement derived from the coupons design iteration. It's also the only view across all empty/error states that shows full-screen, no reason to keep it temporarily for longer if the change has already been approved by design.

> In PointOfSaleIemListEmptyView we've modified the "search" icon for a warning icon and the fonts style (ie: title posOnSurfaceVariantHighest to posOnSurface)

Yes, I'm basing it on the design discussions that we should prevent merchants from interacting with POS if there are no products.

I updated `PointOfSaleIemListEmptyView` to use the same design as `PointOfSaleItemListErrorView`. Since at least for coupons, they both use the same structure now. I will double-check with the design if that's an intention or now.

staskus committed:
Display exclamationMark on the empty view depending on a coupons feature flag

staskus committed:
Make empty view scrollable to support large dynamic type size

staskus committed:
Make PointOfSaleItemListErrorView scrollable to support large dynamic type size

staskus commented:
@iamgabrielma, I will continue displaying an old search icon when the feature flag is disabled.

The intention was to reuse the same empty view for both products and coupons, but I'm clarifying it here (pdfdoF-6PX-p2#comment-7977) to be sure.

I also made some improvements:
-  To make error and empty views scrollable to support large dynamic type sizes.
- Centered variable products empty view state to match other views

staskus committed:
Merge branch 'trunk' into woomob-256-woo-pos-coupons-list-make-empty-error-disabled-loading-state
